prompb/types.proto

```diff
@@ -54,13 +54,58 @@ message Exemplar {
   int64 timestamp = 3;
 }
 
+// A sparse histogram.
+message Histogram {
+  enum ResetHint {
+    UNKNOWN = 0; // Need to test for a counter reset explicitly.
+    YES     = 1; // This is the 1st histogram after a counter reset.
+    NO      = 2; // There was no counter reset between this and the previous Histogram.
+    GAUGE   = 3; // This is a gauge histograms where counter resets don't happen.
+  }
+  
+  uint64 count       = 1;
+  double count_float = 2; // Overrides count if > 0.
+  double sum         = 3;
+  // The schema defines the bucket schema. Currently, valid numbers
+  // are -4 <= n <= 8. They are all for base-2 bucket schemas, where 1
+  // is a bucket boundary in each case, and then each power of two is
+  // divided into 2^n logarithmic buckets. Or in other words, each
+  // bucket boundary is the previous boundary times 2^(2^-n). In the
+  // future, more bucket schemas may be added using numbers < -4 or >
+  // 8.
+  sint32 schema             = 4;
+  double zero_threshold     = 5; // Breadth of the zero bucket.
+  uint64 zero_count         = 6; // Count in zero bucket.
+  double zero_count_float   = 7; // Overrides sb_zero_count if > 0.
+  Buckets negative_buckets  = 8;
+  Buckets positive_buckets  = 9;
+  ResetHint reset_hint      = 10;
+  // timestamp is in ms format, see model/timestamp/timestamp.go for
+  // conversion from time.Time to Prometheus timestamp.
+  int64 timestamp = 11;
+} 
+
+// Sparse buckets.
+message Buckets {
+  message Span {
+    sint32 offset = 1; // Gap to previous span, or starting point for 1st span (which can be negative).
+    uint32 length = 2; // Length of consecutive buckets.
```

Comment:
I note these are optional in `client_model`, and I couldn't figure out why.
If the answer is other than "mistake", please explain both cases.

Reply:
I _believe_ that's mostly historical. The proto spec in prometheus/client_model is using the proto2 syntax. Back then, protobuf had optional and non-optional fields. There were different schools of thought. Some considered `optional` fields harmful, some considered non-optional fields harmful. The Prometheus exposition protobuf was created following the latter school of thought, according to which you declare all fields `optional`, even if they aren't really optional.

proto3 resolved the schism by (arguably) declaring the "make everything `optional`" school as the winner. In proto3, all fields are implicitly optional. If they are missing, the default value is filled in, and (on the API level) you cannot actually distinguish between "unset" and "explicitly specified the value that happens to be the default value". If you really want an "unset" state, you have to declare a field as `repeated` but only care about zero or one repeats.

If the above is inaccurate, please pardon me. My context about protobuf is quite incomplete.